src/mdns/mdns.hpp

```diff
@@ -415,14 +422,18 @@ class Publisher
         kMaxTextEntrySize = 255,
     };
 
+    void OnServiceResolved(const std::string &aType, const DiscoveredInstanceInfo &aInstanceInfo);
+    void OnHostResolved(const std::string &aHostName, const DiscoveredHostInfo &aHostInfo);
+
     PublishServiceHandler mServiceHandler        = nullptr;
     void *                mServiceHandlerContext = nullptr;
 
     PublishHostHandler mHostHandler        = nullptr;
     void *             mHostHandlerContext = nullptr;
 
-    DiscoveredServiceInstanceCallback mDiscoveredServiceInstanceCallback = nullptr;
-    DiscoveredHostCallback            mDiscoveredHostCallback            = nullptr;
+    uint64_t mNextSubscriberId = 1;
+
+    std::map<uint64_t, std::pair<DiscoveredServiceInstanceCallback, DiscoveredHostCallback>> mDiscoveredCallbacks;
```

Comment:
nit: Make a struct to hold `DiscoveredServiceInstanceCallback` and `DiscoveredHostCallback` so it can be more expressive.

Reply:
We may have a major refactoring for mdns modules. 
We can do it when that happens.